src/bit_machine/exec.rs

```diff
@@ -397,16 +397,28 @@ impl BitMachine {
                             unreachable!()
                         };
 
-                    if choice_bit {
-                        let padr_a_b = cmp::max(size_a, size_b) - size_b;
-                        self.fwd(1 + padr_a_b);
-                        call_stack.push(CallStack::Back(1 + padr_a_b));
-                        call_stack.push(CallStack::Goto(right));
-                    } else {
-                        let padl_a_b = cmp::max(size_a, size_b) - size_a;
-                        self.fwd(1 + padl_a_b);
-                        call_stack.push(CallStack::Back(1 + padl_a_b));
-                        call_stack.push(CallStack::Goto(left));
+                    match (&ip.inner, choice_bit) {
+                        (RedeemNodeInner::Case(_, right), true)
+                        | (RedeemNodeInner::AssertR(_, right), true) => {
+                            let padr_a_b = cmp::max(size_a, size_b) - size_b;
+                            self.fwd(1 + padr_a_b);
+                            call_stack.push(CallStack::Back(1 + padr_a_b));
+                            call_stack.push(CallStack::Goto(right));
+                        }
+                        (RedeemNodeInner::Case(left, _), false)
+                        | (RedeemNodeInner::AssertL(left, _), false) => {
+                            let padl_a_b = cmp::max(size_a, size_b) - size_a;
+                            self.fwd(1 + padl_a_b);
+                            call_stack.push(CallStack::Back(1 + padl_a_b));
+                            call_stack.push(CallStack::Goto(left));
+                        }
+                        (RedeemNodeInner::AssertL(_, r_cmr), true) => {
+                            return Err(ExecutionError::ReachedPrunedBranch(*r_cmr))
+                        }
+                        (RedeemNodeInner::AssertR(l_cmr, _), false) => {
+                            return Err(ExecutionError::ReachedPrunedBranch(*l_cmr))
+                        }
+                        _ => unreachable!(),
```

Comment:
I feel like we could reduce code duplication somehow, but let's look into that later.

Reply:
I have a later PR which does reduce it somewhat (though at the expense of having a bunch more precomputed fixed vectors)